packages/next/src/lib/metadata/resolve-metadata.ts

```diff
@@ -248,40 +230,79 @@ export async function collectMetadata(
-        ? metadataExport(parentPromise)
-        : metadataExport
-    const layerMetadataPromise =
-      currentMetadata instanceof Promise
-        ? currentMetadata
-        : Promise.resolve(currentMetadata)
+  let resolvingIndex = 0
+  for (let i = 0; i < metadataItems.length; i++) {
+    const [metadataExport, staticFilesMetadata] = metadataItems[i]
+    let metadata: Metadata | null = null
+    if (typeof metadataExport === 'function') {
+      // call each `generateMetadata function concurrently and stash their resolver
+      generateMetadataResults.push(
+        metadataExport(
+          new Promise((resolve) => {
+            resolvers.push(resolve)
+          })
+        )
+      )
 
-    parentPromise = parentPromise.then((resolved) => {
-      return layerMetadataPromise.then((metadata) => {
-        resolved = deepClone(resolved)
-        merge(resolved, metadata, staticFilesMetadata, {
-          title: resolved.title?.template || null,
-          openGraph: resolved.openGraph?.title?.template || null,
-          twitter: resolved.twitter?.title?.template || null,
-        })
-        return process.env.NODE_ENV === 'development'
-          ? Object.freeze(resolved)
-          : resolved
-      })
-    })
+      const resolveParent = resolvers[resolvingIndex]
+      const generatedMetadata = generateMetadataResults[resolvingIndex]
+      resolvingIndex++
+      // In dev we clone and freeze to prevent relying on mutating resolvedMetadata directly.
+      // In prod we just pass resolvedMetadata through without any copying.
+      const currentResolvedMetadata: ResolvedMetadata =
+        process.env.NODE_ENV === 'development'
+          ? Object.freeze(
+              require('next/dist/compiled/@edge-runtime/primitives/structured-clone').structuredClone(
+                resolvedMetadata
+              )
+            )
+          : resolvedMetadata
+
+      // This resolve should unblock the generateMetadata function if it awaited the parent
+      // argument. If it didn't await the parent argument it might already have a value since it was
+      // called concurrently. Regardless we await the return value before continuing on to the next layer
+      resolveParent(currentResolvedMetadata)
+      metadata =
+        generatedMetadata instanceof Promise
+          ? await generatedMetadata
+          : generatedMetadata
+    } else {
+      metadata = metadataExport
+    }
+
+    // If the layout is the same layer with page, skip the
+    if ((i === 0 || (i !== 0 && i !== metadataItems.length - 2)) && metadata) {
+      resolvedTitle = resolveTitle(metadata?.title, resolvedTitle?.template)
+      resolvedTwitterTitle = resolveTitle(
+        metadata?.twitter?.title,
+        resolvedTwitterTitle?.template
+      )
+      resolvedOpenGraphTitle = resolveTitle(
+        metadata?.openGraph?.title,
+        resolvedOpenGraphTitle?.template
+      )
+    }
+
+    merge(resolvedMetadata, metadata, staticFilesMetadata)
```

Comment:
I would still revert the parts where you extracted title merging and just do this instead. It reads much more clearly what is going on
```suggestion
    /** somewhere above we initialize template to empty values **/
    let templates = {
      title: null,
      openGraph: null,
      twitter: null,
    }
    
    /**
     *
     *
     *
     *
     *
     **/
    
    /* now here where we call merge we update templates after merge if the layer is right */
    merge(resolvedMetadata, metadata, staticFilesMetadata, templates)
    if (i < metadataItems.length - 2) {
      // This is the metadata for a Layout that is an ancestor of the Page (the last two items are always the Layout and Page metadata defined at the leaf route level). We want to update title templates when the Layout is an ancestor of the leaf
      templates = {
          title: resolvedMetadata.title?.template || null,
          openGraph: resolvedMetadata.openGraph?.title?.template || null,
          twitter: resolvedMetadata.twitter?.title?.template || null,
        }
    }
```